### PR TITLE
Hell yeah, it's ANF

### PR DESCRIPTION
**Pull request by SquidDev**
It's like normal normal form, but administrative instead.

### CODE REVIEW

**Comment by zardyh** [comment]
That's a lot of commits.

File: src/Core/Core.hs [hunk 1]
```diff
@@ -152,31 +147,26 @@ instance Pretty a => Pretty (CoStmt a) where
 instance Pretty a => Pretty [CoStmt a] where
   pretty = vcat . map pretty
 
-{-# ANN freeIn "HLint: ignore" #-}
```
Comment (zardyh) [inline src/Core/Core.hs]: Bonus points for removing this

File: src/Core/Lower.hs [hunk 2]
```diff
@@ -110,67 +139,60 @@ lowerAt (Access r k _) ty = do
...
-tup2Rec k b = do
-  b' <- lowerType b
-  pure [(T.pack (show k), b')]
+lowerType :: Type Typed -> CoType
+lowerType t@TyTuple{} = CotyExactRows (tup2Rec 1 t)
+lowerType (TyArr a b) = CotyArr (lowerType a) (lowerType b)
+lowerType (TyForall vs b) = foldr (CotyForall . unTvName) (lowerType b) vs
+lowerType (TyApp a b) = CotyApp (lowerType a) (lowerType b)
+lowerType (TyRows rho vs) = CotyRows (lowerType rho) (map (fmap lowerType) vs)
+lowerType (TyExactRows vs) = CotyExactRows (map (fmap lowerType) vs)
```
Comment (zardyh) [inline src/Core/Lower.hs]: `map (fmap` must be some kind of sin. I'll allow it, though.

File: src/Core/Optimise/Inline.hs [hunk 3]
```diff
@@ -1,103 +1,84 @@
...
-betaReduce :: TransformPass
-betaReduce = pass go where
-  go term = case term of
-    CotApp (CotLam Small (var, tp) body) ex -> do
-      (var', ref) <- invent tp
-      pure . CotLet [(var', tp, ex)] .  substitute (Map.singleton var ref) $ body
-    CotTyApp (CotLam Big (var, _) body) tp ->
-      pure $ substituteInTys (Map.singleton var tp) body
-    _ -> pure term
+inlineVariablePass :: IsVar a => [CoStmt a] -> [CoStmt a]
```
Comment (zardyh) [inline src/Core/Optimise/Inline.hs]: Is this still live?

File: src/Core/Simplify.hs [hunk 4]
```diff
@@ -2,33 +2,26 @@ module Core.Simplify
...
-      , matchKnownConstr
-      , matchOfMatch
-      , matchOfBottom
-      , inlineVariable
-      , betaReduce
-      , dropUselessLet
-      -- , complexSafePropag
-      ]
+optmOnce = pure . passes where
+  passes = foldl (.) id $ reverse
```
Comment (zardyh) [inline src/Core/Simplify.hs]: This is `mconcat` for the Endo monoid

File: src/Core/Optimise/Inline.hs (hunk 3 above)
Comment (SquidDev) [reply src/Core/Optimise/Inline.hs]: Yeah, this now works again. It's horribly slow in GHCi, but it works.

**Pull request merged**